### ISSUE TITLE
Make age parallel

### ISSUE THREAD

**Issue by paulmillr** [FiloSottile/age#109]
If you encrypt files on a machine with tons of RAM and cores, age isn't any faster versus some basic slow PC.

I think it would be great to utilize resources when they're available.

**Comment by RKinsey** [comment]
This is that it isn't feasible to do that without overhauling Go's cryptograhpy libraries (and might be unsafe, I don't know enough about goroutine security to say for sure).   

The only functions in age that actually handle the plaintext are EncryptOAEP/DecryptOAEP from crypto/rsa and Seal/Open from x/crypto/chacha20poly1305, neither of which are parallel. Both _could_ be parallelized, but RSA generally hasn't because it needs a parallel-friendly modular exponentiation function. ChaCha is fairly easy to parallelize, but Go's implementation is handwritten assembly using vector instructions when available (unless you're using a purego build, gccgo, or an uncommon CPU architecture). I have a feeling that probably outperforms a goroutine version, but maybe not.

**Comment by xorhash** [comment]
@RKinsey I'm not sure if this argument actually holds. `internal/stream/stream.go` seems to read and write in chunks of 64&nbsp;KiB (plus 16&nbsp;bytes of Poly1305 tag for each encrypted chunk). Therefore, there's parallelization potential there by queueing up the encryption/decryption of chunks (or multiples of chunks) between cores. Orchestrating the whole thing so that there's no bottleneck when reading or writing is another story though.

**Comment by joonas-fi** [comment]
Yeah @RKinsey was talking about the key-wrapping phase. The actual symmetric stream encryption is where the bulk of Age's work happens (at least on larger file sizes) and it looks like it could be parallelizable.

The stream is divided into fixed-size chunks of 64 kB, and each chunk uses the same encryption key but of course a different nonce. The nonce is calculated based on the chunk number.

**Comment by Tronic** [comment]
Just running chacha20-poly1305 in parallel for a few blocks easily more than doubles the speed. My own tool is written in Python and does 2.2 GB/s encryption and decryption (using 4 threads for chacha, otherwise single-threaded). It is shame that the crypto libraries don't offer threaded implementations of these algorithms.

**Comment by paulmillr** [comment]
@Tronic are you using the latest rage? The speed difference should be minimal right now

**Comment by Tronic** [comment]
@paulmillr rage 0.7.0 on Windows.
